chain/network/src/time.rs

```diff
@@ -0,0 +1,99 @@
+#![allow(dead_code)]
+use once_cell::sync::Lazy;
+use parking_lot::RwLock;
+use std::sync::Arc;
+
+// TODO: consider wrapping these types to prevent interactions
+// with other time libraries, especially the the direct access
+// to the realtime (i.e. not through the Clock).
+pub type Instant = time::Instant;
+pub type Utc = time::OffsetDateTime;
+pub type Duration = time::Duration;
+
+// Instant doesn't have a deterministic contructor,
+// however since Instant is not convertible to an unix timestamp,
+// we can snapshot Instant::now() at the process startup
+// and treat it as a constant. All observable effects will be then
+// deterministic.
+static FAKE_CLOCK_MONO_START: Lazy<Instant> = Lazy::new(Instant::now);
+
+// An arbitrary non-trivial deterministic Utc timestamp.
+const FAKE_CLOCK_UTC_START: Lazy<Utc> = Lazy::new(|| Utc::from_unix_timestamp(89108233).unwrap());
+
+#[derive(Clone)]
+enum ClockInner {
+    Real,
+    Fake(FakeClock),
+}
+
+/// Clock encapsulates a system clock, allowing to replace it
+/// with a fake in tests.
+/// Since system clock is a source of external information,
+/// it has to be replaced with a fake double, if we want our
+/// tests to be deterministic.
+///
+/// This is a reimplementation of primitives/src/time.rs
+/// with a more systematic approach.
+/// TODO: add tests, put it is some reusable package and use
+/// throughout the nearcore codebase.
+#[derive(Clone)]
+pub struct Clock(ClockInner);
+
+impl Clock {
+    pub fn real() -> Clock {
+        Clock(ClockInner::Real)
+    }
+    /// current time according to the monotonic clock
+    pub fn now(&self) -> Instant {
+        match &self.0 {
+            ClockInner::Real => Instant::now(),
+            ClockInner::Fake(ref fake) => fake.now(),
+        }
+    }
+    /// current time according to the system/walltime clock
+    pub fn now_utc(&self) -> Utc {
+        match self.0 {
+            ClockInner::Real => Utc::now_utc(),
+            ClockInner::Fake(ref fake) => fake.now_utc(),
+        }
+    }
+}
+
+struct FakeClockInner {
+    mono: Instant,
+    utc: Utc,
+}
+
+/// TEST-ONLY
+#[derive(Clone)]
+pub struct FakeClock(Arc<RwLock<FakeClockInner>>);
+
+impl FakeClock {
+    pub fn new(utc: Utc) -> Self {
+        Self(Arc::new(RwLock::new(FakeClockInner { utc, mono: *FAKE_CLOCK_MONO_START })))
+    }
+    pub fn now(&self) -> Instant {
+        self.0.read().mono
+    }
+    pub fn now_utc(&self) -> Utc {
+        self.0.read().utc
+    }
+    pub fn clock(&self) -> Clock {
+        Clock(ClockInner::Fake(self.clone()))
+    }
+    pub fn advance(&self, d: Duration) {
+        assert!(d >= Duration::ZERO);
+        let mut c = self.0.write();
+        c.mono += d;
+        c.utc += d;
+    }
+    pub fn set_utc(&self, utc: Utc) {
+        self.0.write().utc = utc;
+    }
+}
```

Comment:
wdym?

Reply:
Something like this: 

```
impl Clock {
  pub fn new() -> (Clock, TimeMachine)
}

impl TimeMachine {
  pub fn set_utc(&mut self, utc: Utc);
}
```

Basically, the single `FakeClock` type has methods for reading and setting time (now_utc/set_utc). I'd imagine that all users of FakeClock would like to either only read time, or only set time.

Reply:
Well, that's not exactly true: if you have a test helper which is emulating passage of time while interacting with the object under test, it will need to both write advance the time and pass it to the object under test for reading. Then you would have to pass these as 2 arguments and the information that they touch the same clock is lost to the type system. Unless you meant just the set_utc - I imagine this one is useful for testing the behavior during NTP synchronization (I've never actually written tests like that), or for testing some corner cases on specific dates.